ivp/solver/algorithm_test.go

```diff
@@ -0,0 +1,113 @@
+package solver_test
+
+import (
+	"fmt"
+	"log"
+	"math"
+	"testing"
+
+	"gonum.org/v1/gonum/ivp"
+	"gonum.org/v1/gonum/ivp/sim"
+	"gonum.org/v1/gonum/ivp/solver"
+	"gonum.org/v1/gonum/mat"
+)
+
+type TestModel struct {
+	*sim.Model
+	solution *sim.Model
+	err      func(h, i float64) float64
+}
+
+var (
+	Quadratic = new(TestModel)
+)
+
+func setup(t *testing.T) {
+	quad, err := sim.NewModel(mat.NewVecDense(2, []float64{0, 0}),
+		nil, func(y []float64, dom float64, x, u []float64) {
+			y[0], y[1] = x[1], 1.
+		}, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	Quadratic.Model = quad
+	quadsol, err := sim.NewModel(mat.NewVecDense(2, []float64{0, 0}),
+		nil, func(y []float64, dom float64, x, u []float64) {
+			y[0], y[1] = dom*dom/2., dom
+		}, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	Quadratic.solution = quadsol
+	Quadratic.err = func(h, i float64) float64 { return math.Pow(h*i, 4) }
+}
+func TestRK4Quadratic(t *testing.T) {
+	setup(t)
+	var slv ivp.Integrator = new(solver.RK4)
+
+	err := slv.Set(0, Quadratic.Model)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	nx, _ := Quadratic.Model.Dims()
+	steps := 10
+	dt := 0.1
+
+	results := make([]float64, nx)
+
+	solmodel := Quadratic.solution
+	soleq, _ := solmodel.Equations()
+	solDims, _ := solmodel.Dims()
+	solution := make([]float64, solDims)
+	for i := 1.; i < float64(steps+1); i++ {
+		dom := i * dt
+		slv.Step(results, dt)
+		soleq(solution, dom, results, nil)
+		for j := range results {
+			got := math.Abs(solution[j] - results[j])
+			expected := Quadratic.err(dt, i)
+			if got > expected {
+				t.Errorf("error %e greater than threshold %e", got, expected)
+			}
+
+		}
+	}
+}
+
+func Example_fallingBall() {
+	const (
```

Comment:
Example tests are for teaching, so this needs more commentary to lead the user through how to use the package.